Large dnode pool feature

nedbass opened the pull request:
This feature adds support for variable length dnodes. Our motivation is
to eliminate the overhead associated with using spill blocks.  Spill
blocks are used to store system attribute data (i.e. file metadata) that
does not fit in the dnode's bonus buffer. By allowing a larger bonus
buffer area the use of a spill block can be avoided.  Spill blocks
potentially incur an additional read I/O for every dnode in a dnode
block. As a worst case example, reading 32 dnodes from a 16k dnode block
and all of the spill blocks could issue 33 separate reads. Now suppose
those dnodes have size 1024 and therefore don't need spill blocks.  Then
the worst case number of blocks read is reduced to from 33 to two--one
per dnode block. In practice spill blocks may tend to be co-located on
disk with the dnode blocks so the reduction in I/O would not be this
drastic. In a badly fragmented pool, however, the improvement could be
significant.

ZFS-on-Linux systems that make heavy use of extended attributes would
benefit from this feature. In particular, ZFS-on-Linux supports the
xattr=sa dataset property which allows file extended attribute data
to be stored in the dnode bonus buffer as an alternative to the
traditional directory-based format. Workloads such as SELinux and the
Lustre distributed filesystem often store enough xattr data to force
spill bocks when xattr=sa is in effect. Large dnodes may therefore
provide a performance benefit to such systems.

Other use cases that may benefit from this feature include files with
large ACLs and symbolic links with long target names. Furthermore,
this feature may be desirable on other platforms in case future
applications or features are developed that could make use of a
larger bonus buffer area.

adilger commented:
Ned, have you done any performance testing on this patch with Lustre or other xattr-heavy workload?

I can imagine that increasing the dnode size itself would reduce performance to some extent, since more IO is needed for a given number of dnodes.  Once xattrs are included into the test mix, and a real world workload where the dnode attributes and/or xattr are modified after initial creation the performance would swing in favor of the large dnodes due to avoiding seeking for the spill blocks.

jxiong commented:
On my testing node, I noticed that once I enabled this feature, the xattr is not stored as SA any more.

[root@centos7 testpool]# ~/srcs/zfs/zfs/cmd/zfs/zfs get all zfstest/testpool |egrep 'dnodesize|xattr'
zfstest/testpool  xattr                 sa                     local
zfstest/testpool  dnodesize             1K                     local

touch a file and set xattr:

[root@centos7 testpool]# rm -rf tf
[root@centos7 testpool]# touch tf
[root@centos7 testpool]# setfattr -n user.val -v "The size of a dnode may be a multiple of 512 bytes up to the size of" tf
[root@centos7 testpool]# stat tf
  File: ‘tf’
  Size: 0         	Blocks: 1          IO Block: 131072 regular empty file
Device: 26h/38d	Inode: 15          Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Context: unconfined_u:object_r:unlabeled_t:s0
Access: 2015-09-04 10:47:39.241747282 -0400
Modify: 2015-09-04 10:47:39.241747282 -0400
Change: 2015-09-04 10:47:39.241747282 -0400
 Birth: -

And then I found that user.val is actually stored in a separate object:

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
        17    1    16K    512      0      1K    512  100.00  ZFS directory
                                               168   bonus  System attributes
	dnode flags: USED_BYTES USERUSED_ACCOUNTED 
	dnode maxblkid: 0
	path	/tf/<xattrdir>
	uid     0
	gid     0
	atime	Fri Sep  4 10:47:39 2015
	mtime	Fri Sep  4 10:47:42 2015
	ctime	Fri Sep  4 10:47:42 2015
	crtime	Fri Sep  4 10:47:39 2015
	gen	658
	mode	41777
	size	4
	parent	15
	links	2
	pflags	40800000145
	microzap: 512 bytes, 2 entries

		user.val = 21 (type: Regular File)
		security.selinux = 19 (type: Regular File)
Indirect blocks:
               0 L0 EMBEDDED et=0 200L/53P B=659

Is this feature exclusive with SA based xattr, or I just missed something in my setup?

behlendorf commented:
@jxiong enabling the pool feature and setting both the setting __xattr=sa__ and __dnodesize=N__ properties should be sufficient.  Could you post a test script or an example commands for how you ran the test and we'll try and reproduce the issue.  I know Ned tested this pretty extensively before posting anything for review.  But on the surface it sure looks like you did everything right.

jxiong commented:
@behlendorf the commands I used to do the test are regular ones, just create a testpool with xattr set to sa and dnodesize set to 1K. I was trying to see how large lustre layout can be stored in 1K dnode. The commands are as follows:

[root@centos7 zfs]# zpool create -f zfstest raidz2 /Users/jxiong/vdevs/d{0..9}
[root@centos7 zfs]# zfs create -o xattr=sa -o dnodesize=1k zfstest/testpool
[root@centos7 zfs]# zfs get all zfstest/testpool |egrep 'dnodesize|xattr'
zfstest/testpool  xattr                 sa                     local
zfstest/testpool  dnodesize             1K                     local
[root@centos7 zfs]# touch /zfstest/testpool/tf
[root@centos7 zfs]# setfattr -n user.val -v "helloworkd" /zfstest/testpool/tf
[root@centos7 zfs]# stat --printf="%i\n" /zfstest/testpool/tf
7
[root@centos7 zfs]# zdb -vvvv zfstest/testpool 7
....

Before the patch is applied, I should be able to see that user.val is stored in bonus buffer, but now I can see the user.val is stored in another object just like 'xattr=on' wasn't taken effect.

[root@centos7 zfs]# zdb -vvvv zfstest/testpool 11
Dataset zfstest/testpool [ZPL], ID 71, cr_txg 5, 37.1K, 9 objects, rootbp DVA[0]=<0:16800:600> DVA[1]=<0:2e00de00:600> [L0 DMU objset] fletcher4 lz4 LE contiguous unique double size=800L/200P birth=89L/89P fill=9 cksum=b30a661a1:44c445301f7:d9f4cd414a58:1db06447b7fe03

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
        11    1    16K    512     1K      1K    512  100.00  ZFS plain file
                                               168   bonus  System attributes
	dnode flags: USED_BYTES USERUSED_ACCOUNTED 
	dnode maxblkid: 0
	path	/tf/<xattrdir>/user.val
	uid     0
	gid     0
	atime	Wed Sep  9 18:04:27 2015
	mtime	Wed Sep  9 18:04:27 2015
	ctime	Wed Sep  9 18:04:27 2015
	crtime	Wed Sep  9 18:04:27 2015
	gen	89
	mode	100644
	size	10
	parent	9
	links	1
	pflags	40800000005
Indirect blocks:
               0 L0 0:12c00:600 200L/200P F=1 B=89/89

		segment [0000000000000000, 0000000000000200) size   512

nedbass commented:
@jxiong I'm unable to reproduce your results.  It is working as expected for me.  Can you post the git commit revision you are testing, and confirm that your user space tools match the running kernel?

Below is a transcript of me running your test case. You can see from the zdb output that the xattrs are stored in the bonus buffer as expected.

```
$  bass6@firstdown ~/zfs /dev/pts/0 Sun Sep 13 16:20:17 (large_dnode_feat u=) >
git log --pretty=oneline | head -1
0da0e790a28488e8cf1e6ea6a1c0a6640a6f0b15 Implement large_dnode pool feature
$  bass6@firstdown ~/zfs /dev/pts/0 Sun Sep 13 16:20:25 (large_dnode_feat u=) >
sudo ./cmd/zpool/zpool create -f zfstest raidz2 /home/bass6/vdevs/d{0..9}
$  bass6@firstdown ~/zfs /dev/pts/0 Sun Sep 13 16:20:57 (large_dnode_feat u=) >
sudo ./cmd/zfs/zfs create -o xattr=sa -o dnodesize=1k zfstest/testpool
$  bass6@firstdown ~/zfs /dev/pts/0 Sun Sep 13 16:21:03 (large_dnode_feat u=) >
sudo ./cmd/zfs/zfs get all zfstest/testpool |egrep 'dnodesize|xattr'
zfstest/testpool  xattr                 sa                     local
zfstest/testpool  dnodesize             1K                     local
$  bass6@firstdown ~/zfs /dev/pts/0 Sun Sep 13 16:21:14 (large_dnode_feat u=) >
sudo touch /zfstest/testpool/tf
$  bass6@firstdown ~/zfs /dev/pts/0 Sun Sep 13 16:21:31 (large_dnode_feat u=) >
sudo  setfattr -n user.val -v "helloworkd" /zfstest/testpool/tf
$  bass6@firstdown ~/zfs /dev/pts/0 Sun Sep 13 16:21:45 (large_dnode_feat u=) >
stat --printf="%i\n" /zfstest/testpool/tf
7
$  bass6@firstdown ~/zfs /dev/pts/0 Sun Sep 13 16:21:59 (large_dnode_feat u=) >
sudo ./cmd/zdb/zdb -vvvv zfstest/testpool 7
Dataset zfstest/testpool [ZPL], ID 71, cr_txg 11, 35.9K, 7 objects, rootbp DVA[0]=<0:cc00:600> DVA[1]=<0:f00cc00:600> [L0 DMU objset] fletcher4 lz4 LE contiguous unique double size=800L/200P birth=21L/21P fill=7 cksum=d11c61e75:50fe17f8e3a:102dc32f47fb3:23860111b46c19

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         7    1    16K    512      0      1K    512    0.00  ZFS plain file
                                               236   bonus  System attributes
	dnode flags: USERUSED_ACCOUNTED 
	dnode maxblkid: 0
	path	/tf
	uid     0
	gid     0
	atime	Sun Sep 13 16:21:31 2015
	mtime	Sun Sep 13 16:21:31 2015
	ctime	Sun Sep 13 16:21:31 2015
	crtime	Sun Sep 13 16:21:31 2015
	gen	18
	mode	100644
	size	0
	parent	4
	links	1
	pflags	40800000004
	SA xattrs: 60 bytes, 1 entries

		user.val = helloworkd
Indirect blocks:


$  bass6@firstdown ~/zfs /dev/pts/0 Sun Sep 13 16:22:06 (large_dnode_feat u=) >
sudo ./cmd/zdb/zdb -vvvv zfstest/testpool 9
Dataset zfstest/testpool [ZPL], ID 71, cr_txg 11, 35.9K, 7 objects, rootbp DVA[0]=<0:cc00:600> DVA[1]=<0:f00cc00:600> [L0 DMU objset] fletcher4 lz4 LE contiguous unique double size=800L/200P birth=21L/21P fill=7 cksum=d11c61e75:50fe17f8e3a:102dc32f47fb3:23860111b46c19

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
zdb: dmu_bonus_hold(9) failed, errno 2
$  bass6@firstdown ~/zfs /dev/pts/0 Sun Sep 13 16:22:18 (large_dnode_feat u=) >
sudo ./cmd/zdb/zdb -vvvv zfstest/testpool 11
Dataset zfstest/testpool [ZPL], ID 71, cr_txg 11, 35.9K, 7 objects, rootbp DVA[0]=<0:cc00:600> DVA[1]=<0:f00cc00:600> [L0 DMU objset] fletcher4 lz4 LE contiguous unique double size=800L/200P birth=21L/21P fill=7 cksum=d11c61e75:50fe17f8e3a:102dc32f47fb3:23860111b46c19

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
zdb: dmu_bonus_hold(11) failed, errno 2
```

nedbass commented:
I can reproduce the problem after rebasing the patch on master. Thanks for the bug report, I'll run down the root cause.

nedbass commented:
@jxiong I've identified the cause of the issue you reported. Pull request #3788 should address it.

jxiong commented:
@nedbass I will try it out. Thanks a lot.

behlendorf commented:
@ahrens the proposed large dnode feature is ready for review.  Can you please look this patch over.  The only major missing bit of functionality missing is send/recv support to pools which don't have large dnode support.  What are your thoughts on this?

I've created a new 'OpenZFS Review' tag so we can more easily track these reviews going forward.

https://github.com/zfsonlinux/zfs/labels/OpenZFS%20Review

ahrens commented on include/sys/dnode.h:
```diff
@@ -202,6 +221,7 @@ typedef struct dnode {
 	uint32_t dn_datablksz;		/* in bytes */
 	uint64_t dn_maxblkid;
 	uint8_t dn_next_type[TXG_SIZE];
+	uint8_t dn_count;		/* metadnode slots consumed on disk */
```
good that there's a comment explaining what this is, but even better would be to name the variable more evocatively.  e.g. dn_num_slots

ahrens commented on include/sys/dmu.h:
```diff
@@ -754,7 +755,8 @@ typedef struct dmu_object_info {
 	uint8_t doi_compress;
 	uint8_t doi_nblkptr;
-	uint8_t doi_pad[4];
+	uint8_t doi_count;
```
count of what?  see comment about dn_count.  consider renaming, or at least add a comment here.

ahrens commented on module/zfs/zfeature_common.c:
```diff
@@ -242,4 +242,15 @@ zpool_feature_init(void)
+		SPA_FEATURE_EXTENSIBLE_DATASET,
+		SPA_FEATURE_NONE
+	};
+	zfeature_register(SPA_FEATURE_LARGE_DNODE,
```
ZoL should sync up with https://github.com/illumos/illumos-gate/commit/ca0cc3918a1789fa839194af2a9245f801a06b1a (5959 clean up per-dataset feature count code), and then you should make this a ZFEATURE_FLAG_PER_DATASET.  That should clean up a bunch of your feature flag handling code.

vozhyk- replied:
@ahrens 
Thanks for mentioning this. It will also be useful for the lz4hc feature.

behlendorf replied:
Definitely, we'll want to get this merged on the ZoL side first.

vozhyk- replied:
@behlendorf 
I'm trying to port this right now. If I succeed, I'll open a pull request.

vozhyk- replied:
By the way, the pull request for the port is #3965.

ahrens commented on module/zfs/dmu_send.c:
```diff
@@ -1537,7 +1541,8 @@ restore_object(struct restorearg *ra, objset_t *os, struct drr_object *drro)
 
-	tx = dmu_tx_create(os);
+	count = DN_BONUSLEN_TO_COUNT(drro->drr_bonuslen);
+	tx = dmu_tx_create_dn_count(os, count);
```
OK so you can create an object with a specific slot count (just like you can create an object with a specific type, bonus length, block size, etc).  I think that you should specify the object parameters when you allocate it, with dmu_object_alloc/claim/reclaim, rather than having it "figure it out" from the tx.

If the tx also needs to know the slot count, that's fine.  Ideally it would be specified with dmu_tx_hold_*().

behlendorf replied:
@ahrens my understanding is that what your suggesting is what implemented in the initial version of the patch.  However, that ended up being problematic because it's possible that the preferred dnode size could change mid-way through a transaction.  Adding it to the tx seemed like the cleanest way to ensure the same dnode size was used all the way through the tx.  This does seem like it should be done in dmu_tx_hold_*().

ahrens replied:
Why does it matter if the preferred dnode size changes mid-way through the transaction?  I think the ZPL should decide what it's going to do, and then tell the DMU with dmu_tx_hold_* and dmu_object_alloc().

behlendorf replied:
I've explained poorly.  Right, it doesn't matter as long we've safely stashed it away with dmu_tx_hold_() for consistent use latter.  So I think we're in completely agreement.  I was just saying the initial prototype patch ran in to problems because it naively pulled the size from the property as needed.  That ended up causing some ASSERTs when it changed at the wrong time.  I'm sure @nedbass could have done a better job explaining.  :)

ahrens commented on module/zfs/dmu_send.c:
```diff
@@ -1598,18 +1603,30 @@ restore_freeobjects(struct restorearg *ra, objset_t *os,
+	if (obj == 0)
+		obj++;
+
+	while (obj < drrfo->drr_firstobj + drrfo->drr_numobjs) {
```
Performance of this will be very bad for certain streams, because you are not using dmu_object_next() to advance to the next allocated object.

behlendorf replied:
Yes, we should look in to how to handle this better.

ahrens commented:
Can you remind me why you need to be able to change the dnode size after the fact, as opposed to setting it only at filesystem creation time?  Seems like that would be simpler so I'd like to understand the use case.

ahrens commented:
Having zil replay create the object with a different size seems wrong to me.  What if a subsequent log record could require the additional space in the bonus buffer?  It seems like the ZIL code is assuming that the bonus buffer is only used by the SA code which can (I presume) dynamically deal with any bonus buffer size.

I would suggest you find a way to store the num_slots in the lr_create_t (or at worst, add a new log record type).  For example, you could use some of the high 16 bits of lr_foid, which is currently at most DN_MAX_OBJECT.  (use the BF64_* macros to carve it up into a bitfield).

adilger commented:
On Oct 23, 2015, at 17:44, Matthew Ahrens <notifications@github.com> wrote:
> Can you remind me why you need to be able to change the dnode size after the fact, as opposed to setting it only at filesystem creation time? Seems like that would be simpler so I'd like to understand the use case.
> 

I think the main benefit is that the administrator doesn't need to guess in advance what dnode size to select, not whether that size is appropriate or optimal for all users/apps in the filesystem. Different usage patterns may create different amounts of extended metadata (inline xattrs/SAs) that are difficult to know in advance. 

Another major benefit of allowing dynamic dnode sizes is that this would make the feature immediately useful for existing datasets rather than only new ones.

behlendorf commented:
> Can you remind me why you need to be able to change the dnode size after the fact.

Making the dnode size fixed at dataset create time would have definitely simplify things.  Although I think it would also make the feature far less useful for exactly the reasons @adilger mentions.

I know internally we have large existing filesystems (with ~3+ billion files) which we rely on 24/7.  We'd benefit greatly from being able to enable this feature on an existing dataset and have all new dnodes properly sized.  Even better I could see this feature being enhanced so existing dnodes could be grown/shrunk in place if the needed slots were available.  That's a huge win for existing filesystems.

> Having zil replay create the object with a different size seems wrong to me.

This is another area where @nedbass and I went back and forth.  If at all possible we wanted to avoid adding a new log record type and we were leery of squatting on any existing fields.  That led us to deciding that large dnodes were just an optimization and that in the unlikely case of a crash creating a few of non-optimally sized dnodes wouldn't be the worst thing.  That said, if you're OK with us using the upper bits of lr_foid I like that much better.

@ahrens do you have any ideas about how send/recv should be handled to a pool which doesn't support this feature?  Can we just punt on it and avoid handling this case (my preference), or must we be excessively clever and generate a stream which is backwards compatible on the fly?

Thanks for the review!  I'm sure @nedbass will be able to chime in the next week or so.

ahrens commented:
> That led us to deciding that large dnodes were just an optimization and that in the unlikely case of a crash creating a few of non-optimally sized dnodes wouldn't be the worst thing. That said, if you're OK with us using the upper bits of lr_foid I like that much better.

@behlendorf I think if you wanted to make large dnodes truly and optimization, it would need to be tightly integrated with the SA code and not usable any other way.  Contrast with the spill block, which is only used the SA (I think) but it works correctly without the SA and it would be easy to correctly add a different consumer.

Given that large dnodes is are already changing the on-disk format, I don't see why modifying the ZIL format (either to use upper bits of lr_foid or to add a new log record type) is a big deal.

> do you have any ideas about how send/recv should be handled to a pool which doesn't support this feature?

The recv should gracefully fail.  You can handle this like the spill block, by adding a new DRR_BACKUP_FEATURE_*.

behlendorf commented:
Thanks for the feedback and pointer to DMU_BACKUP_FEATURE_* flags.  We'll try and get it refreshed in the next few weeks.

nedbass commented:
I've updated the branch to address the feedback from @ahrens. In particular, the following has changed:

* Increased large dnode test coverage in `ztest`:
 - Pick a random dnode size for each new object.
 - Fill unused bonus buffer space with non-zero data
 - Add a test that
    - Visits each object in a dataset
    - Verifies the dnode size matches what was specified at create time
    - Verifies the unused bonus buffer space is not corrupted
* Eliminated storage of the dnode size in the `dmu_tx`.  Instead, add new DMU interfaces to allow consumers to specify the dnode size for new objects.
* ZIL replay now creates objects with the correct dnode size.
* Use DMU_BACKUP_FEATURE_* flag so `zfs recv` will gracefully fail with an incompatible pool.

nedbass commented:
Refreshed. Fixed cstyle issues and fixed hang in ztest due to unreleased dbuf holds.

nedbass commented:
Incorporated review comments from @ahrens. 

* Simplified `dnode_is_{free,allocated,consumed}()` function prototypes.
* Removed redundant checking of dnode type in `dnode_hold_impl()`

bzzz77 commented:
just tried to benchmark:

clean zfs/master:
Created 1000000 in 29414ms in 1 threads - 33997/sec
Created 1000000 in 20045ms in 2 threads - 49887/sec
Created 1000000 in 19259ms in 4 threads - 51923/sec
Created 1000000 in 17284ms in 8 threads - 57856/sec

zfs/master + large dnodes:
Created 1000000 in 40618ms in 1 threads - 24619/sec
Created 1000000 in 28142ms in 2 threads - 35534/sec
Created 1000000 in 25731ms in 4 threads - 38863/sec
Created 1000000 in 25244ms in 8 threads - 39613/sec

is this expected behavior ?

nedbass commented:
@bzzz77 thanks for the benchmark results. The performance degradation is not expected. Can you provide more detail about your test setup?

Also, you must set the `dnodesize` dataset property change the default dnode size.

bzzz77 commented:
well, this is a small box with iSCSI over GigE (~110MB/s) running createmany from Lustre tree. I didn't change anything since zpool create and verified with zdb that dnode size is still 512bytes.

nedbass commented:
@bzzz77 I updated the patch to address the performance regression. With large dnodes `dmu_object_next()` needs to call `dmu_object_info()` to figure out how many dnodes to skip in the current block.  It turns out this introduces a significant throughput penalty on create rates. To fix this I updated `dmu_object_alloc()` to avoid the call to `dmu_object_next()` altogether, and instead simply retry the allocation starting from the next dnode block. This approach seems to be significantly faster, even for master without the large dnodes patch. I plan to submit that change as a separate pull request for master to illicit some discussion and testing.

bzzz77 commented:
I can confirm this patch is doing much better.

adilger commented on module/zfs/dmu.c:
```diff
@@ -1929,6 +1930,18 @@ dmu_object_size_from_db(dmu_buf_t *db_fake, uint32_t *blksize,
 }
```
Looks like dmu_object_size_from_db() also needs to be updated.  The "add 1 for dnode space" is incorrect for large dnodes, so this should be:

        /* add in the number of slots used for the dnode itself */
        *nbk512 = ... SPA_MINBLOCKSHIFT) + (dn->dn_num_slots << DNODE_SHIFT);

nedbass replied:
Good catch.

nedbass commented:
Refreshed patch rebased on #4711.  Other changes:

* Further reworked metadnode backfill algorithm to avoid potential pathology with large dnodes. A low `blk_fill` count could make full dnode blocks wappear sparse.  The allocator would then repeatedly attempt and fail to allocate from these full blocks.  For example, a full block with 4k dnodes would appear to be 25% full according to the blk_fill field, so it would be selected for backfill. To mitigate this the code now switches to searching for empty blocks after too many failures.

* Changed the allowed values for the `dnodesize` property from a numeric value to `legacy | auto`. This avoids burdening users with having to figure out a reasonable number. Currently `auto` just sets a default size of 1k, which is a good choice for most common xattr workloads (e.g. Lustre, SELinux). Future code improvements could dynamically choose a dnode size based on observed workloads.

adilger commented:
While I see the "auto" and "legacy" options as the best useful useful for most users, I don't think that allowing a numerical value is overly complex. Users don't have to specify the dnode size if they don't want, but this could be very useful for benchmarking and performance tuning in the future. That is doubly true because this is a new feature and we don't necessarily have a good idea of what the optimum value is for different workloads. For example, Samba makes heavy use of xattrs and may benefit from larger dnodes than other configurations.

As for the fill count vs. large dnode size, this could negate some of the gains being made on the file creation front if there is a mixed create/unlink workload since it means that more full blocks may be reused before their time...

Does it make sense to keep an in-memory counter of dnode sizes and dnode counts to work out the running average dnode size (possibly taking into account the _desired_ dnode size even when there wasn't enough space for it).  The counters could be per-CPU and do not need to be perfectly uptodate, so long as they are consistent in the sum and count   

That would help in both of these cases.  It would give the expansion factor for the full count, to determine when a dnode leaf block should be considered full. It would also help the "auto" case know how large to allocate dnodes rather than having a fixed size and assuming that is best.

nedbass commented:
@adilger thanks for the feedback. I just refreshed the patch with a version that allows some literal values to be specified.  For simplicity of the UI I restricted it to powers of 2 from 1k to 16k.  I was having trouble coming up with a UI I was happy with that allowed both integer and string property values.  The existing ZFS property interfaces don't support that in a flexible way.  So I settled on limiting allowed values to `legacy | auto | 1k | 2k | 4k | 8k | 16k`.  Ideally I would have liked to support integer and shorthand formats for input and output, and to support arbitrary multiples of 512, but it didn't seem worth the code complexity that would entail. So just exposing the shorthand powers-of-2 seemed like a reasonable compromise.

nedbass commented:
Refreshed with test cases added to ZFS Test Suite.  Also rebased to include some improvements to xattrtest which the test suite takes advantage of.

behlendorf commented:
Thanks everybody, merged to master:

50c957f Implement large_dnode pool feature
68cbd56 Backfill metadnode more intelligently
8128558 xattrtest: allow verify with -R and other improvements
